includes/forms.php

```diff
@@ -0,0 +1,640 @@
+	}
+
+	ob_start();
+
+	if ( isset( $_REQUEST['rc_action'] ) && $_REQUEST['rc_action'] === "lostpassword" ) {
+
+		echo rc_lost_password_form();
+
+	} elseif ( isset( $_REQUEST['rc_action'] ) && $_REQUEST['rc_action'] === "lostpassword_checkemail" ) {
+
+		echo rc_lost_password_check_email_message();
+
+	} elseif ( isset( $_REQUEST['rc_action'] ) && ( $_REQUEST['rc_action'] === "lostpassword_reset" || $_REQUEST['rc_action'] === "reset-password" ) ) {
+
+		echo rc_change_password_form( $args );
+
+	} else {
+
+		if ( ! is_user_logged_in() ) :
+
+			rc_show_error_messages( 'login' );
+
+			if ( isset( $_GET['password-reset'] ) && $_GET['password-reset'] == 'true' ) : ?>
+				<div class="rc-message">
+					<p class="rc-success"><?php _e( 'Password changed successfully. Please log in with your new password.', 'restrict-content' ); ?></p>
+				</div>
+			<?php endif; ?>
+
+			<form id="rc_login_form"  class="<?php echo esc_attr( $args['class'] ); ?>" method="POST" action="<?php echo esc_url( rc_get_current_url() ); ?>">
+				<fieldset class="rc_login_data">
+
+					<p>
+						<label for="rc_user_login"><?php _e( 'Username', 'restrict-content' ); ?></label>
+						<input name="rc_user_login" id="rc_user_login" class="required" type="text" />
+					</p>
+
+					<p>
+						<label for="rc_user_pass"><?php _e( 'Password', 'restrict-content' ); ?></label>
+						<input name="rc_user_pass" id="rc_user_pass" class="required" type="password"/>
+					</p>
+
+					<p>
+						<input type="checkbox" name="rc_user_remember" id="rc_user_remember" value="1"/>
+						<label for="rc_user_remember"><?php _e( 'Remember me', 'restrict-content' ); ?></label>
+					</p>
+
+					<p class="rc_lost_password">
+						<a href="<?php echo esc_url( add_query_arg( 'rc_action', 'lostpassword') ); ?>">
+							<?php _e( 'Lost your password?', 'restrict-content' ); ?>
+						</a>
+					</p>
+
+					<p>
+						<input type="hidden" name="rc_action" value="login"/>
+						<input type="hidden" name="rc_redirect" value="<?php echo esc_url( $args['redirect'] ); ?>"/>
+						<input type="hidden" name="rc_login_nonce" value="<?php echo wp_create_nonce( 'rc-login-nonce' ); ?>"/>
+						<input id="rc_login_submit" type="submit" value="<?php esc_attr_e( 'Login', 'restrict-content' ); ?>"/>
+					</p>
+
+				</fieldset>
+			</form>
+
+		<?php else : ?>
+
+			<div class="rc_logged_in"><a href="<?php echo wp_logout_url( home_url() ); ?>"><?php _e( 'Log out', 'restrict-content' ); ?></a></div>
+
+		<?php endif;
+
+	}
+
+	return ob_get_clean();
+}
+
+/**
+ * Processes the login form
+ */
+function rc_process_login_form() {
+
+	if ( ! isset( $_POST['rc_action'] ) || 'login' !== $_POST['rc_action'] ) {
+		return;
+	}
+
+	if ( ! isset( $_POST['rc_login_nonce'] ) || ! wp_verify_nonce( $_POST['rc_login_nonce'], 'rc-login-nonce' ) ) {
+		return;
+	}
+
+	if ( is_email( $_POST['rc_user_login'] ) && ! username_exists( $_POST['rc_user_login'] ) ) {
+
+		$user = get_user_by( 'email', $_POST['rc_user_login'] );
```

Comment:
Can we please sanitize this and other raw $_POST vars?

Reply:
@pippinsplugins I didn't because they get sanitized deeper in the stack, but we can do it again here if you want.